--- xfce4-mixer-4.3.90.1/configure.ac.orig	2006-04-16 16:58:22.000000000 +0200
+++ xfce4-mixer-4.3.90.1/configure.ac	2006-04-27 22:39:08.982941500 +0200
@@ -83,7 +83,7 @@
 
 dnl Check for i18n support
 XDT_I18N([ar az be bg bn_IN ca cs de el en_GB es es_MX et eu fa fi fr gu
-          he hi hu hy it ja ko lt mr ms nl nb_NO pa pl pt_BR pt_PT ro ru
+          he hi hu hy it ja ko lt mr ms nl nb pa pl pt_BR pt ro ru
 	  sk sv ta tr uk vi zh_CN zh_TW])
 
 AC_MSG_CHECKING([whether to build big mixer window])
