--- xfce4-mixer-4.1.99.3/configure.ac.orig	2004-12-20 22:32:25.000000000 +0100
+++ xfce4-mixer-4.1.99.3/configure.ac	2004-12-28 11:28:19.134018368 +0100
@@ -68,7 +68,7 @@
 
 dnl Check for i18n support
 BM_I18N([xfce4-mixer], [ar az be bg ca de en_GB es es_MX et eu fa fi fr gu he
-			hu it ja ko lt ms nl nb_NO pa pt_BR pt_PT ro ru sk tr vi
+			hu it ja ko lt ms nl nb pa pt_BR pt ro ru sk tr vi
 			zh_CN zh_TW])
 
 BM_DEPEND([XML], [libxml-2.0], [2.4.0])
