--- xfce4-mixer-4.1.99.1/configure.ac~	2004-11-06 15:43:24.000000000 +0100
+++ xfce4-mixer-4.1.99.1/configure.ac	2004-11-25 07:33:15.770327432 +0100
@@ -55,7 +55,7 @@
 
 dnl Check for i18n support
 BM_I18N([xfce4-mixer], [ar az be bg ca de es es_MX et eu fa fi fr gu he hu it
-			ja ko lt ms nl no pa pt_BR pt_PT ru sk tr vi zh_CN 
+			ja ko lt ms nl nb pa pt_BR pt ru sk tr vi zh_CN 
 			zh_TW])
 
 BM_DEPEND([XML], [libxml-2.0], [2.4.0])
